// Node service definitions for Master-Worker gRPC communication
// Requirements: 1.4, 14.1

syntax = "proto3";

package antcode.v1;

import "common.proto";

// NodeService handles all communication between Master and Worker nodes
service NodeService {
  // Bidirectional streaming for real-time communication
  rpc NodeStream(stream NodeMessage) returns (stream MasterMessage);
  
  // Single call for node registration
  rpc Register(RegisterRequest) returns (RegisterResponse);
}

// Messages sent from Worker to Master
message NodeMessage {
  oneof payload {
    Heartbeat heartbeat = 1;
    LogBatch log_batch = 2;
    TaskStatus task_status = 3;
    TaskAck task_ack = 4;
    CancelAck cancel_ack = 5;
  }
}

// Messages sent from Master to Worker
message MasterMessage {
  oneof payload {
    TaskDispatch task_dispatch = 1;
    TaskCancel task_cancel = 2;
    ConfigUpdate config_update = 3;
    Ping ping = 4;
  }
}


// Heartbeat message sent periodically by Worker to Master
message Heartbeat {
  string node_id = 1;                      // Unique identifier for the node
  string status = 2;                       // Node status (online, busy, offline)
  Metrics metrics = 3;                     // System metrics
  OSInfo os_info = 4;                      // Operating system information
  Timestamp timestamp = 5;                 // Message timestamp
  map<string, string> capabilities = 6;   // Node capabilities
}

// Batch of log entries sent from Worker to Master
message LogBatch {
  repeated LogEntry logs = 1;    // List of log entries
  bool compressed = 2;           // Whether compressed_data is used
  bytes compressed_data = 3;     // Gzip compressed log data (if compressed=true)
}

// Single log entry
message LogEntry {
  string execution_id = 1;   // Task execution identifier
  string log_type = 2;       // Log type (stdout, stderr, system)
  string content = 3;        // Log content
  Timestamp timestamp = 4;   // Log timestamp
}

// Task execution status update from Worker to Master
message TaskStatus {
  string execution_id = 1;           // Task execution identifier
  string status = 2;                 // Status (pending, running, success, failed, cancelled, timeout)
  optional int32 exit_code = 3;      // Process exit code (if success/failed/timeout/cancelled)
  optional string error_message = 4; // Error message (if failed)
  Timestamp timestamp = 5;           // Status update timestamp
}

// Task dispatch from Master to Worker
message TaskDispatch {
  string task_id = 1;                      // Unique task identifier
  string project_id = 2;                   // Project identifier
  string project_type = 3;                 // Project type (spider, code, file)
  int32 priority = 4;                      // Task priority (higher = more urgent)
  map<string, string> params = 5;          // Task parameters
  map<string, string> environment = 6;     // Environment variables
  int32 timeout = 7;                       // Execution timeout in seconds
  string download_url = 8;                 // Project download URL
  string file_hash = 9;                    // Project file hash for verification
  string entry_point = 10;                 // Entry point script/file
}

// Task acknowledgment from Worker to Master
message TaskAck {
  string task_id = 1;           // Task identifier being acknowledged
  bool accepted = 2;            // Whether task was accepted
  optional string reason = 3;   // Rejection reason (if not accepted)
}

// Task cancellation request from Master to Worker
message TaskCancel {
  string task_id = 1;        // Task identifier to cancel
  string execution_id = 2;   // Execution identifier to cancel
}

// Cancellation acknowledgment from Worker to Master
message CancelAck {
  string task_id = 1;           // Task identifier
  bool success = 2;             // Whether cancellation succeeded
  optional string reason = 3;   // Failure reason (if not successful)
}

// Node registration request
message RegisterRequest {
  string machine_code = 1;                 // Unique machine identifier
  string api_key = 2;                      // API key for authentication
  string node_id = 3;                      // Node identifier
  OSInfo os_info = 4;                      // Operating system information
  map<string, string> capabilities = 5;   // Node capabilities
}

// Node registration response
message RegisterResponse {
  bool success = 1;                  // Whether registration succeeded
  string node_id = 2;                // Assigned node identifier
  optional string error = 3;         // Error message (if failed)
  int32 heartbeat_interval = 4;      // Recommended heartbeat interval in seconds
}

// Configuration update from Master to Worker
message ConfigUpdate {
  map<string, string> config = 1;   // Configuration key-value pairs
}

// Ping message for keepalive
message Ping {
  Timestamp timestamp = 1;   // Ping timestamp
}
